### PR TITLE
add action input as parameters for tool execution in conversational agent

### PR DESCRIPTION
**Pull request by jngz-es**
### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
https://github.com/opensearch-project/ml-commons/issues/3134

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java [hunk 1]
```diff
@@ -472,6 +472,11 @@ public static Map<String, String> constructToolParams(
         if (toolSpecConfigMap != null) {
             toolParams.putAll(toolSpecConfigMap);
         }
+        toolParams.put("action_input", actionInput);
+        if (isJson(actionInput)) {
```
[the remark this answers is not in the document]
Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java:476]: Yes.

[the remark this answers is not in the document]
Comment (dbwiddis) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java:476]: Good catch, @pyek-bot looking at `MLToolSpec` we'd get here:
https://github.com/opensearch-project/ml-commons/blob/5cfdc3ce9aa89d79d0ada175d1a497e1942fd75a/common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java#L109-L118

and [`Gson.toJson`](https://www.javadoc.io/static/com.google.code.gson/gson/2.8.5/com/google/gson/Gson.html#toJson-java.lang.Object-) internally uses `Object.getClass()` which may throw NPE.

**Review by mingshl — DISMISSED** [review]

**Comment by mingshl** [comment]
tests passed but failed in upload, not related to this code change. Approved. 
```
Run actions/upload-artifact@v4
/usr/bin/docker exec  a6c6ef6ad38d2e9993d03ba2bdc50f2146f892a4a109fc9fecaf2c66802943f0 sh -c "cat /etc/*release | grep ^ID"
/__e/node[20](https://github.com/opensearch-project/ml-commons/actions/runs/11670434010/job/32954019963?pr=3200#step:8:21)/bin/node: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /__e/node20/bin/node)
/__e/node20/bin/node: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by /__e/node20/bin/node)
``'

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java [hunk 2]
```diff
@@ -472,6 +472,11 @@ public static Map<String, String> constructToolParams(
         if (toolSpecConfigMap != null) {
             toolParams.putAll(toolSpecConfigMap);
         }
+        toolParams.put("action_input", actionInput);
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java:475]: `action_input` looks like coupled with code logic. Change to more readable name ?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java:476]: @jngz-es Can you add a test case for `null` value ?

**Comment by reuschling** [comment]
The changes for AgentUtils looks fine, but is AgentUtils used for conversational agents? The tool parameters are build inside `MLConversationalFlowAgentRunner.getToolExecuteParams(MLToolSpec toolSpec, Map<String, String> params)`, i.e. [here](https://github.com/opensearch-project/ml-commons/blob/9c1b8a8c1e28ab8bdaf507cb71799c4809c3d8c6/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLConversationalFlowAgentRunner.java#L411). There is no AgentUtils invocation. The only invocation of `AgentUtils.‎‎constructToolParams` I found is [inside MLChatAgentRunner](https://github.com/opensearch-project/ml-commons/blob/9c1b8a8c1e28ab8bdaf507cb71799c4809c3d8c6/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java#L287).

I showed a code proposal for `getToolExecuteParams` at https://github.com/opensearch-project/ml-commons/pull/2977#discussion_r1806572463, the only difference is that there is no dedicated actionInput parameter, the actionInput is the "input" entry so far, which has to be temporarily stored inside a local variable.

I would also highly recommend to add the new "action_input" parameter to flow agents (i.e. `MLFlowAgentRunner.getToolExecuteParams`) also. There is no use of AgentUtils too. Of course there is the possibility with `parameters.previous_tool_name.output`, but tool specifications should act the same independent where they should be used, whether inside flow or conversational agents.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java (hunk 2 above)
Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java]: Changed.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java]: Added.

**Comment by reuschling** [comment]
> @reuschling Thanks for reviewing this PR. From [#2918 (comment)](https://github.com/opensearch-project/ml-commons/issues/2918#issuecomment-2340233740), I think you know "MLChatAgentRunner uses the AgentUtils method [AgentUtils.constructToolParams](https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/AgentUtils.java#L458) for generating the params for a tool."
> 
> The name seems confusing, but actually `MLChatAgentRunner` is for conversational agent ([code link](https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLAgentExecutor.java#L310)). 

Thanks a lot @ylwu-amzn for clarification with the code link. Yes the name confused me, sorry about that @jngz-es . In this case I'm also fine with the changes :)

**Review by ylwu-amzn — APPROVED** [review]

**Review by b4sjoo — APPROVED** [review]
